src/mastic.rs

```diff
@@ -0,0 +1,331 @@
+    pub fn new(
+        algorithm_id: u32,
+        szk: Szk<T, P, SEED_SIZE>,
+        vpf: Vidpf<VidpfWeight<T::Field>, 16>,
+        bits: usize,
+    ) -> Self {
+        Self {
+            algorithm_id,
+            szk,
+            vpf,
+            bits,
+            phantom: PhantomData,
+        }
+    }
+}
+/// Mastic aggregation parameter.
+///
+/// This includes the VIDPF tree level under evaluation, a set of prefixes to evaluate at that level,
+/// and, optionally, the aggregate results of prior levels.
+pub type MasticAggregationParam = Poplar1AggregationParam;
+
+/// Contains broadcast information shared between parties to support VIDPF correctness.
+pub type MasticPublicShare<V> = VidpfPublicShare<V>;
+
+/// Add necessary traits for MasticPublicShare here
+
+/// Message sent by the [`Client`] to each Aggregator during the Sharding phase.
+#[derive(Clone, Debug)]
+pub struct MasticInputShare<F: FieldElement, const SEED_SIZE: usize> {
+    /// VIDPF key share.
+    vidpf_key: VidpfKey,
+
+    /// The proof share.
+    proofs_share: SzkProofShare<F, SEED_SIZE>,
+}
+impl<'a, T, P, const SEED_SIZE: usize> ParameterizedDecode<(&'a Mastic<T, P, SEED_SIZE>, usize)>
+    for MasticInputShare<T::Field, SEED_SIZE>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    fn decode_with_param(
+        decoding_parameter: &(&'a Mastic<T, P, SEED_SIZE>, usize),
+        bytes: &mut Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        let vidpf_key = VidpfKey::decode(bytes)?;
+        let proofs_share =
+            SzkProofShare::<T::Field, SEED_SIZE>::decode_with_param(decoding_parameter.0, bytes)?;
+        Ok(Self {
+            vidpf_key,
+            proofs_share,
+        })
+    }
+}
+
+impl<F: FieldElement, const SEED_SIZE: usize> Encode for MasticInputShare<F, SEED_SIZE> {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        self.vidpf_key.encode(bytes)?;
+        self.proofs_share.encode(bytes)?;
+        Ok(())
+    }
+
+    fn encoded_len(&self) -> Option<usize> {
+        Some(self.vidpf_key.encoded_len()? + self.proofs_share.encoded_len()?)
+    }
+}
+/// Struct containing a vector of VIDPF outputs: one for each prefix. Likely needs to be changed in the next PR to include proofs as well.
+#[derive(Clone, Debug)]
+pub struct MasticOutputShare<V: VidpfValue> {
+    result: Vec<V>,
+}
+impl<V: VidpfValue + Debug> Aggregatable for MasticOutputShare<V> {
+    type OutputShare = MasticOutputShare<V>;
+    /// Update an aggregate share by merging it with another (`agg_share`).
+    fn merge(&mut self, agg_share: &Self) -> Result<(), VdafError> {
+        if self.result.len() != agg_share.result.len() {
+            return Err(VdafError::Uncategorized(
+                "Attempted to merge two output shares with different agg_params".to_string(),
+            ));
+        };
+        // Would love to get rid of the below clone if possible.
```

Comment:
I think we would have to add a `for<'a> Add<&'a Self, Output = Self>` bound to the `IdpfValue` trait to do so.